chapter01/src/main/java/com/seaofnodes/simple/parser/AST.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * DO NOT REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+ * <p>
+ * Contributor(s):
+ * <p>
+ * The Initial Developer of the Original Software is Dibyendu Majumdar.
+ * <p>
+ * This file is part of the Sea Of Nodes Simple Programming language
+ * implementation. See https://github.com/SeaOfNodes/Simple
+ * <p>
+ * The contents of this file are subject to the terms of the
+ * Apache License Version 2 (the "APL"). You may not use this
+ * file except in compliance with the License. A copy of the
+ * APL may be obtained from:
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package com.seaofnodes.simple.parser;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * A simple AST definition.
+ */
+public class AST {
```

Comment:
IMHO, a ASTree should be a ... "Tree".  Should have kids and a count thereof; kids should be generic so a simple visitor can walk the tree.  Basically I'm asking for a field  `AST[] kids` which replaces the corresponding children parts in the binary ops.  This array can be either null or 0-length if you have no kids (but pick one style).  The various other ASTs can add named accessors, like "lhs()" and "rhs()" for 2-entry binary ops, or "pred()", "if_t()" and "if_f()" for IfASTs.